harness/determined/pytorch/_lightning/adapter.py

```diff
@@ -0,0 +1,209 @@
+        for freq in opt_frequencies:
+            check.eq(freq, 1, "custom optimizer frequencies are not supported")
+        optimizers = cast(List[Optimizer], optimizers)
+        lr_scheduler_dicts = cast(List[dict], lr_scheduler_dicts)
+
+        def lightning_scheduler_dict_to_det(lrs: dict) -> _LRScheduler:
+            """
+            input_dict = {
+                'scheduler': None,
+                'name': None,  # no custom name
+                'interval': 'epoch',  # after epoch is over
+                'frequency': 1,  # every epoch/batch
+                'reduce_on_plateau': False,  # most often not ReduceLROnPlateau scheduler
+                'monitor': monitor,  # value to monitor for ReduceLROnPlateau
+                'strict': True,  # enforce that the monitor exists for ReduceLROnPlateau
+            }
+            """
+            # TODO support custom frequencies with the manual step.
+            step_mode = (
+                LRScheduler.StepMode.STEP_EVERY_EPOCH
+                if lrs["interval"] == "epoch"
+                else LRScheduler.StepMode.STEP_EVERY_BATCH
+            )
+            return context.wrap_lr_scheduler(lrs["scheduler"], step_mode)
+
+        optimizers = [self.context.wrap_optimizer(opt) for opt in optimizers]
+        lr_schedulers = [lightning_scheduler_dict_to_det(lrs) for lrs in lr_scheduler_dicts]
+        return optimizers, lr_schedulers
+
+    def train_batch(
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Union[torch.Tensor, Dict[str, Any]]:
+        self.lm.on_train_batch_start(batch, batch_idx, dataloader_idx=0)
+        type(self.lm).global_step = mock.PropertyMock(return_value=batch_idx)
+
+        opt_metrics = []
+
+        _optimizers = self.optimizers
+
+        def get_optimizers(self, *args, **kwargs) -> List[Optimizer]:
+            return _optimizers
+
+        for opt_idx, opt in enumerate(self.optimizers):
+            with monkey_patch.monkey_patch(self.lm, "optimizers", get_optimizers):
+                self.lm.toggle_optimizer(opt, opt_idx)
+            metrics = self.lm.training_step(batch, batch_idx, optimizer_idx=opt_idx)
+
+            self.lm.on_train_batch_end(metrics, batch, batch_idx, dataloader_idx=0)
+
+            if metrics is None:
+                continue
+            elif not isinstance(metrics, dict):
+                metrics = {"loss": metrics}
+
+            self.context.backward(metrics["loss"])
+            self.lm.on_after_backward()
+            self.context.step_optimizer(opt, on_before_zero_grad=self.lm.on_before_zero_grad)
+
+            opt_metrics.append(metrics)
+            with monkey_patch.monkey_patch(self.lm, "optimizers", get_optimizers):
+                self.lm.untoggle_optimizer(opt_idx)
+
+        agg_metrics = {}
+        for opt_idx, rv in enumerate(opt_metrics):
+            for k, v in rv.items():
+                agg_metrics[f"opt{opt_idx}_{k}"] = v
```

Comment:
I think it's wrong to add prefixes here since users are the ones to control the names. If they specify the same names for different optimizers, it should be wrong unless PytorchLightning itself implements in this way

Reply:
Actually, I am wrong. You need to add prefixes.

Reply:
I think we can probably keep the user's names unless they collide with each other. always adding prefixes was an easier first approach.